internal/statesync/stateprovider.go

```diff
@@ -355,76 +354,113 @@ func (s *stateProviderP2P) addProvider(p lightprovider.Provider) {
 // consensusParams sends out a request for consensus params blocking
 // until one is returned.
 //
-// If it fails to get a valid set of consensus params from any of the
-// providers it returns an error; however, it will retry indefinitely
-// (with backoff) until the context is canceled.
-func (s *stateProviderP2P) consensusParams(ctx context.Context, height int64) (types.ConsensusParams, error) {
-	var iterCount int64
-
-	timer := time.NewTimer(0)
-	defer timer.Stop()
-	for {
-		params, err := s.tryGetConsensusParamsFromWitnesses(ctx, height)
-		if err != nil {
-			return types.ConsensusParams{}, fmt.Errorf("problem getting parameters from witnesses: %w", err)
+// It attempts to send requests to all witnesses in parallel, but if
+// non responds it will retry them all sometime later until it
+// receives some response. This operation will block until it receives
+// a response or the context is canceled.
+func (s *stateProviderP2P) consensusParams(ctx context.Context, height int64) (param types.ConsensusParams, err error) {
+	defer func() {
+		// it simplifies some of the concurrency handling if
```

Comment:
No objection to the recovery per se, but I don't think this will work: You can only recover panics from the same goroutine that calls panic, cf. https://go.dev/play/p/W7t5mnxwVIp for example.

Reply:
ah good catch. (actually my real response is "booooo") will fix this up.